sound/pci/hda/hda_codec.c

```diff
@@ -3281,6 +3281,36 @@ int snd_hda_codec_parse_pcms(struct hda_codec *codec)
 }
 EXPORT_SYMBOL_GPL(snd_hda_codec_parse_pcms);
 
+/* assign all PCMs of the given codec */
+int snd_hda_codec_build_pcms_noattach(struct hda_codec *codec)
+{
+	struct hda_bus *bus = codec->bus;
+	struct hda_pcm *cpcm;
+	int dev, err;
+
+	err = snd_hda_codec_parse_pcms(codec);
+	if (err < 0)
+		return err;
+
+	/* attach a new PCM streams */
+	list_for_each_entry(cpcm, &codec->pcm_list_head, list) {
+		if (cpcm->pcm)
+			continue; /* already attached */
+		if (!cpcm->stream[0].substreams && !cpcm->stream[1].substreams)
+			continue; /* no substreams assigned */
+
+		dev = get_empty_pcm_device(bus, cpcm->pcm_type);
+		if (dev < 0) {
+			cpcm->device = SNDRV_PCM_INVALID_DEVICE;
+			continue; /* no fatal error */
+		}
+		cpcm->device = dev;
```

Comment:
I am lost here.
For the ASoC driver, the PCMs are created by topology, and the back-ends are typically based on hardware.
Are you suggesting that we have PCM devices visible to the user, and an implicit 1:1 connection to the HDMI back-end? That would preclude e.g. any sort of processing/volume control on such outputs.

Reply:
@plbossart Above is verbatim code copied from patch_hdmi.c, just dropping the attachment part (thus the _noattach() postfix to function name).

The way it works is that the actual PCMs are created via topology as normal. In our HDAC wrapper, we will pass some information from asoc/topology world into the legacy patch_hdmi.c.  One such information is mapping of iDisp port iDisp# to a PCM device number. This could in theory be an arbitrary routing in topology, but unfortunately patch_hdmi.c does offer a way to configure the numbers from outside, rather it's fixed inside.

Additionally, there appears to be some legacy who expect a certain fixed PCM numbering. I'm not 100% convinced how valid this still is, but definitely this is way HDMI is by most current users.

I'll prototype the change to actually pass the PCM numbers to patch_hdmi.c dynamically. We can then discuss how disruptive this is to other users of the HDA code, and whether the legacy user concern is valid.